File: domain/research_map.py
"""研究地图 Schema 定义。

围绕结构化知识卡，定义研究地图最小可用实体、关系、时间线、聚类与研究空白结构。
"""

from typing import Any, Dict, List, Literal

from pydantic import BaseModel, Field


EntityType = Literal["paper", "topic", "problem", "method", "dataset", "metric", "gap"]
RelationType = Literal[
    "studies_topic",
    "addresses_problem",
    "uses_method",
    "evaluates_on_dataset",
    "evaluated_by_metric",
    "related_to",
    "evolves_from",
    "highlights_gap",
]


class ResearchEntitySchema(BaseModel):
    """研究地图实体节点。"""

    id: str = Field(..., description="全局唯一节点 ID")
    entity_type: EntityType = Field(..., description="实体类型")
    label: str = Field(..., description="展示名称")
    aliases: List[str] = Field(default_factory=list, description="别名或原始候选名")
    paper_ids: List[str] = Field(default_factory=list, description="关联论文 ID 列表")
    mentions: int = Field(default=0, description="在样本中出现的次数")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="附加元数据")


class ResearchRelationSchema(BaseModel):
    """研究地图边。"""

    source: str = Field(..., description="源节点 ID")
    target: str = Field(..., description="目标节点 ID")
    relation_type: RelationType = Field(..., description="关系类型")
    weight: float = Field(default=1.0, description="关系强度")
    evidence_paper_ids: List[str] = Field(default_factory=list, description="支持该关系的论文 ID")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="附加信息，如共享实体、年份跨度")


class ResearchTimelineEventSchema(BaseModel):
    """年度时间线事件。"""

    year: str = Field(default="unknown", description="年份")
    paper_ids: List[str] = Field(default_factory=list, description="该年份论文 ID")
    key_topics: List[str] = Field(default_factory=list, description="该年份高频主题")
    key_methods: List[str] = Field(default_factory=list, description="该年份高频方法")
    highlights: List[str] = Field(default_factory=list, description="年度摘要")


class ResearchClusterSchema(BaseModel):
    """研究簇。"""

    id: str = Field(..., description="聚类 ID")
    label: str = Field(..., description="簇标签")
    paper_ids: List[str] = Field(default_factory=list, description="簇内论文 ID")
    topic_ids: List[str] = Field(default_factory=list, description="簇覆盖主题 ID")
    method_ids: List[str] = Field(default_factory=list, description="簇覆盖方法 ID")
    summary: str = Field(default="", description="簇摘要")


class ResearchGapSchema(BaseModel):
    """研究空白或候选机会点。"""

    id: str = Field(..., description="空白 ID")
    title: str = Field(..., description="空白标题")
    description: str = Field(default="", description="空白说明")
    gap_type: Literal["explicit_gap", "sparse_topic", "missing_evaluation"] = Field(
        ..., description="空白类型"
    )
    priority: Literal["high", "medium", "low"] = Field(default="medium", description="优先级")
    evidence_paper_ids: List[str] = Field(default_factory=list, description="支持空白的论文 ID")
    related_entity_ids: List[str] = Field(default_factory=list, description="相关主题/方法/指标等节点 ID")


class ResearchMapSchema(BaseModel):
    """研究地图整体输出。"""

    overview: Dict[str, Any] = Field(default_factory=dict, description="总体统计信息")
    entities: List[ResearchEntitySchema] = Field(default_factory=list, description="全部节点")
    relations: List[ResearchRelationSchema] = Field(default_factory=list, description="全部边")
    timeline: List[ResearchTimelineEventSchema] = Field(default_factory=list, description="时间线")
    clusters: List[ResearchClusterSchema] = Field(default_factory=list, description="研究簇")
    gaps: List[ResearchGapSchema] = Field(default_factory=list, description="研究空白")

File: vue-frontend/src/types/research-map.ts
/** 研究地图类型定义 */

export type EntityType =
  | 'paper'
  | 'topic'
  | 'problem'
  | 'method'
  | 'dataset'
  | 'metric'
  | 'gap'

export type RelationType =
  | 'studies_topic'
  | 'addresses_problem'
  | 'uses_method'
  | 'evaluates_on_dataset'
  | 'evaluated_by_metric'
  | 'related_to'
  | 'evolves_from'
  | 'highlights_gap'

export interface ResearchEntity {
  id: string
  entity_type: EntityType
  label: string
  aliases: string[]
  paper_ids: string[]
  mentions: number
  metadata: Record<string, any>
}

export interface ResearchRelation {
  source: string
  target: string
  relation_type: RelationType
  weight: number
  evidence_paper_ids: string[]
  metadata: Record<string, any>
}

export interface TimelineEvent {
  year: string
  paper_ids: string[]
  key_topics: string[]
  key_methods: string[]
  highlights: string[]
}

export interface ResearchCluster {
  id: string
  label: string
  paper_ids: string[]
  topic_ids: string[]
  method_ids: string[]
  summary: string
}

export interface ResearchGap {
  id: string
  title: string
  description: string
  gap_type: 'explicit_gap' | 'sparse_topic' | 'missing_evaluation'
  priority: 'high' | 'medium' | 'low'
  evidence_paper_ids: string[]
  related_entity_ids: string[]
}

export interface ResearchMapData {
  overview: Record<string, any>
  entities: ResearchEntity[]
  relations: ResearchRelation[]
  timeline: TimelineEvent[]
  clusters: ResearchCluster[]
  gaps: ResearchGap[]
}

/** 图谱节点（用于前端可视化） */
export interface GraphNode {
  id: string
  entity_type: EntityType
  label: string
  x?: number
  y?: number
  vx?: number
  vy?: number
  r?: number // 半径（圆形节点）
  width?: number // 宽度（矩形节点）
  height?: number // 高度（矩形节点）
  color: string
  size: number
  paper_ids: string[]
  mentions: number
  metadata: Record<string, any>
  selected?: boolean
  highlighted?: boolean
}

/** 图谱边（用于前端可视化） */
export interface GraphEdge {
  id: string
  source: string | GraphNode
  target: string | GraphNode
  relation_type: RelationType
  weight: number
  label: string
  color: string
  width: number
  paper_ids: string[]
  metadata: Record<string, any>
  selected?: boolean
  highlighted?: boolean
}

/** 图谱视图状态 */
export interface GraphViewState {
  transform: {
    x: number
    y: number
    k: number
  }
  selectedNode: GraphNode | null
  selectedEdge: GraphEdge | null
  hoveredNode: GraphNode | null
  hoveredEdge: GraphEdge | null
  filters: {
    entityTypes: EntityType[]
    minMentions: number
  }
  layout: 'force' | 'circular' | 'hierarchical'
}

/** 实体类型颜色配置 */
export const ENTITY_TYPE_COLORS: Record<EntityType, string> = {
  paper: '#3b82f6',       // 蓝色
  topic: '#10b981',       // 绿色
  problem: '#f59e0b',     // 橙色
  method: '#8b5cf6',      // 紫色
  dataset: '#ec4899',     // 粉色
  metric: '#06b6d4',      // 青色
  gap: '#ef4444',         // 红色
}

/** 关系类型标签映射 */
export const RELATION_LABELS: Record<RelationType, string> = {
  studies_topic: '研究主题',
  addresses_problem: '解决问题',
  uses_method: '使用方法',
  evaluates_on_dataset: '评估数据集',
  evaluated_by_metric: '评估指标',
  related_to: '相关',
  evolves_from: '演化自',
  highlights_gap: '指出空白',
}

/** 实体类型标签映射 */
export const ENTITY_TYPE_LABELS: Record<EntityType, string> = {
  paper: '论文',
  topic: '主题',
  problem: '问题',
  method: '方法',
  dataset: '数据集',
  metric: '指标',
  gap: '空白',
}
